logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/containsHelpers.kt

```diff
@@ -46,7 +50,14 @@ internal fun <E : Any> createExplanatoryAssertionGroup(
             //TODO 0.18.0 looks a lot like toBeNullIfNullGiven
             if (assertionCreatorOrNull != null) {
                 // we don't use a subject, we will not show it anyway
-                it.collectAssertions(container, None, assertionCreatorOrNull)
+                val collectingExpect = CollectingExpect<E>(None, container.components)
+                // not using addAssertionsCreatedBy on purpose so that we don't append a failing assertion
+                collectingExpect.assertionCreatorOrNull()
+                if(collectingExpect.getAssertions().isEmpty()) {
+                    it.collectAssertions(container, None, assertionCreatorOrNull).failing
```

Comment:
Would it be better to make `collectAssertions(...)` fail the explanatory group when the assertionCreator doesn't create any assertions?

Reply:
sounds like a reasonable thing, that an explanatory group can fail is rather a new thing

Reply:
I realize I'm not actually sure how to modify the builder to accomplish this. 

I'm looking at `AssertionOptionExplanatoryExtensions.kt` and I'm not sure how to get the extension method `.collectAssertions` to produce a failing final step.

My naive attempt:
```kotlin
fun <T, G : ExplanatoryAssertionGroupType, R> AssertionsOption<G, R>.collectAssertions(
    container: AssertionContainer<*>,
    maybeSubject: Option<T>,
    assertionCreator: Expect<T>.() -> Unit
): R {
    val collectingExpect = CollectingExpect<T>(None, container.components)
    // not using addAssertionsCreatedBy on purpose so that we don't append a failing assertion
    collectingExpect.assertionCreator()
    return withAssertions(container.collectForCompositionBasedOnSubject(maybeSubject, assertionCreator))
        .let {
            if(collectingExpect.getAssertions().isEmpty()) it.failing
            else it
        }
}
```
has issues because the return type is the generic type `R` and it doesn't recognize `.failing`. I'm not really understanding here how the builder pattern is implemented here with the generic `AssertionsOption<G,R>` type.

Reply:
You need to restrict R with an upper bound of ExplanatoryGroup.FinalStep.kt in AssertionOptionExplanatoryExtensions.kt.

Reply:
With the above type signature the compiler complained about two things:
* It expects the method to return `R` but the inferred type is `ExplanatoryGroup.FinalStep`
* Elsewhere in the code, in other places, it complains about the upper bound `R : ExplanatoryGroup.FinalStep`, saying `ExplanatoryAssertionGroupFinalStep` isn't a subtype of `ExplanatoryGroup.FinalStep`.

I had to use the following to satisfy the compiler.
```kotlin
fun <T, G : ExplanatoryAssertionGroupType, R: ExplanatoryAssertionGroupFinalStep> AssertionsOption<G, R>.collectAssertions(
    container: AssertionContainer<*>,
    maybeSubject: Option<T>,
    assertionCreator: Expect<T>.() -> Unit
): ExplanatoryAssertionGroupFinalStep {
  // ...
}
```